lib/apple/pub_key.go

```diff
@@ -0,0 +1,50 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apple
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"fmt"
+	"math/big"
+)
+
+// ECDSAPublicKeyFromRaw reads an ECDSA public key from a raw Apple public key,
+// as returned by SecKeyCopyExternalRepresentation.
+func ECDSAPublicKeyFromRaw(pubKeyRaw []byte) (*ecdsa.PublicKey, error) {
+	// Verify key length to avoid a potential panic below.
+	// 3 is the smallest number that clears it, but in practice 65 is the more
+	// common length.
+	// Apple's docs make no guarantees, hence no assumptions are made here.
+	if len(pubKeyRaw) < 3 {
+		return nil, fmt.Errorf("public key representation too small (%v bytes)", len(pubKeyRaw))
+	}
+
+	// "For an elliptic curve public key, the format follows the ANSI X9.63
+	// standard using a byte string of 04 || X || Y. (...) All of these
+	// representations use constant size integers, including leading zeros as
+	// needed."
+	// https://developer.apple.com/documentation/security/1643698-seckeycopyexternalrepresentation?language=objc
+	pubKeyRaw = pubKeyRaw[1:] // skip 0x04
+	l := len(pubKeyRaw) / 2
+	x := pubKeyRaw[:l]
+	y := pubKeyRaw[l:]
+
+	return &ecdsa.PublicKey{
+		Curve: elliptic.P256(),
+		X:     (&big.Int{}).SetBytes(x),
+		Y:     (&big.Int{}).SetBytes(y),
+	}, nil
+}
```

Comment:
This was moved from lib/auth/touchid, no code changes. I've added a new test exclusively for it - touchid _does_ test this aspect, but it seemed appropriate to test here, given that it's now public.

Reply:
Yeah, take that $previous_author, this is mine now.  :P